Preserve certain state properties on some sites during reload and navigation

daniel-wer opened the pull request:
This PR aims to preserve certain state properties, like a search query or a drop down selection, during a page reload and back/forward browser navigation.
I've implemented this for a couple of our admin sites and wanted to request feedback first, before adding it to all of the lists :)

This PR also adds libdefs for react-router and libdef stubs for Antd. Before this we always lost prop type information of components when using the `withRouter` or `Form.create` wrappers.
I've also removed the libdefs for backbone and marionette which brought a fascinating speed-up for flows initial type checking (down from >1min to ~10s :tada:), although we didn't even use them.

If you agree with the method I used to preserve those state properties, I'll continue to add this to the other admin pages.

The issue also demands a refresh button for our lists. I'm not sure whether that's worth it, considering that a browser reload is pretty fast. What do you think?

### Mailable description of changes:
 - Search criteria of all lists are now preserved during page reloads and forward/backward browser navigation.

### Steps to test:
- Open the task list and select a user and press search or/and enter a search term. Reloading the page should keep the entered selections/search terms. Navigating somewhere else and using the back button should as well.
- Same for the search queries in the dataset and projects lists.
- In the explorative tracings list, it should additionally be remembered whether archived or open annotations are being displayed.

### Issues:
- fixes #2265

------
- [x] Ready for review

hotzenklotz commented on app/assets/javascripts/app.js:
```diff
@@ -2,12 +2,12 @@
 
 class OxalisApplication {
-  oxalis: ?OxalisController;
+  oxalis: ?typeof OxalisController;
```
There is no longer a reference to `app.oxalis` anywhere. Hence, we can remove this type.

hotzenklotz commented on app/assets/javascripts/components/secured_route.js:
```diff
@@ -47,15 +48,21 @@ class SecuredRoute extends React.PureComponent<Props, State> {
+            } else if (render != null) {
+              return render(props);
+            } else {
+              throw Error("Did specify neither component nor render function.");
```
"Did specify" --> "Specified"

hotzenklotz commented on app/assets/javascripts/dashboard/views/explorative_annotations_view.js:
```diff
@@ -83,15 +90,16 @@ class ExplorativeAnnotationsView extends React.PureComponent<Props, State> {
-
-    const isFinishedString = this.state.shouldShowArchivedTracings.toString();
+    // Cache shouldShowArchivedTracings, otherwise it could have another value later
+    const showArchivedTracings = this.state.shouldShowArchivedTracings;
```
Nice catch!

hotzenklotz commented on app/assets/javascripts/admin/views/project/project_list_view.js:
```diff
@@ -82,11 +83,19 @@ class ProjectListView extends React.PureComponent<StateProps, State> {
     });
   };
 
+  _setState = (...args) => this.setState(...args);
```
This is repeated in every component using the `StatePersistenceComponent`. Perhaps we can find a clever way to include this into it directly.

hotzenklotz commented on app/assets/javascripts/admin/views/task/task_search_form.js:
```diff
@@ -45,6 +56,13 @@ class TaskSearchForm extends React.Component<Props, State> {
     }
   }
 
+  componentDidUpdate(prevProps, prevState) {
```
I am not quite sure what this code does. :-(

philippotto commented:
We should keep in mind that this PR allows state to be persisted across different code versions. Strictly speaking, we would need something similar to the mongo evolutions. As soon as we change code which affects persisted state (e.g., the name of a state attribute or a data structure), this will horribly break, since the state is automatically made invalid. We should think about a way to avoid this.

philippotto commented:
> Navigating somewhere else and using the back button should as well.

This does not work for me when setting a user in the tasks page :(

daniel-wer commented:
> This does not work for me when setting a user in the tasks page :(
You probably forgot to press "Search" after selecting the user - only then the search results are updated and the selection is persisted :)

daniel-wer commented:
@philippotto I've addressed your concerns as we discussed :)
The HOC was replaced with two functions, that make the state changes much more explicit, you were absolutely right about the magic aspect of the previous solution!
I've also forked the prop-types module to allow it to run in production. I'm using it to make sure that the persisted state that is restored to the component state has the correct data types and doesn't break the component :)
Also without the HOC, flow now works correctly and complains if a property is selected to be persisted that is not part of the components state!

If you have a little bit of time over the weekend and could have a quick look, I would appreciate that and finish this PR up by adding this functionality to every admin page and by creating new snaphots.

philippotto commented on app/assets/javascripts/components/state_persistence_component.js:
```diff
@@ -1,55 +1,55 @@
 // @flow
```
Now, the path / file name does not seem right, since its content is not a react component anymore. Maybe move it to libs?

philippotto commented on app/assets/javascripts/components/state_persistence_component.js:
```diff
@@ -1,55 +1,55 @@
-        break;
-      }
+  name: string,
+  stateProperties: { [$Keys<T>]: Function },
```
Nice, didn't know that flow understands `[$Keys<T>]`.

philippotto commented on app/assets/javascripts/admin/views/project/project_list_view.js:
```diff
@@ -26,23 +27,39 @@ type StateProps = {
 };
 
-class ProjectListView extends React.PureComponent<StateProps, State> {
+const STATE_TO_BE_PERSISTED: { [$Keys<State>]: Function } = {
```
Since you are planning to use this in several places, we might want to try to make this even more concise. How about the following as an interface?

```
const persistence: Persistence<State> = new Persistence({
    searchQuery: PropTypes.string,
}, "projectList");

persistence.load(this.props.history);

persistence.persist(this.props.history, nextState);
```

The class could look something like [this](0C4TwDgpgBAysCGxoF4oG8BQUoEsDOAMgPbwAmOAdgOYBcUARkUQDYTwUA0WUSewdfAE6UqXAL4BuDBgDGzeHjxQAChEF58SCjIgAeGAD503AJAV4AWwgDgw6lJN9EEACpEAQhFXrNEUnQB5egArCBlgKWwMExkiCiEAV3CiQQAKaMcEJDdPbw0).

daniel-wer replied:
@philippotto Yep, I like that :) I cannot see your suggestion for the class though, Github shows a Not Found page :/

philippotto replied:
Sorry about that; I fixed the link in the upper comment now :)

philippotto commented on app/assets/javascripts/components/state_persistence_component.js:
```diff
@@ -1,55 +1,55 @@
+    try {
+      // Check whether the type of the persisted state conforms to that of the component to avoid messing up
+      // the components state - this could happen if the type of a state property changed
+      PropTypes.checkPropTypes(stateProperties, persistedState, "persisted state property", name);
```
Very cool that this is so easy to use!

philippotto commented:
> @philippotto I've addressed your concerns as we discussed :)
The HOC was replaced with two functions, that make the state changes much more explicit, you were absolutely right about the magic aspect of the previous solution!
I've also forked the prop-types module to allow it to run in production. I'm using it to make sure that the persisted state that is restored to the component state has the correct data types and doesn't break the component :)

Very cool! Should be a lot more stable regarding future changes now. The code looks nice! I commented a suggestion on how the persistence interface could be a bit more concise. No strong feelings here, though :)

> Also without the HOC, flow now works correctly and complains if a property is selected to be persisted that is not part of the components state!

Awesome!

A random thought: Could the persistence break if a component is used in different places (since it's rather generic)? For example, the task list might be used in the dashboard but also in the admin view. If persistence is used there, could it happen that the dashboard shows state from the admin page? Maybe we can enforce somehow that this cannot happen?

daniel-wer commented:
@philippotto The persistence will not break if a component is used in different places as the location state is tied to a path (`history.replace(path, [state])`), see https://github.com/ReactTraining/history#navigation :)

philippotto commented:
> @philippotto The persistence will not break if a component is used in different places as the location state is tied to a path (history.replace(path, [state])), see ReactTraining/history#navigation :)

Awesome :)

daniel-wer commented:
I've added persistence to the obvious admin view candidates, I think this is sufficient for now :)
If you LGTM the last commits, I'd merge.

philippotto commented:
> I've added persistence to the obvious admin view candidates, I think this is sufficient for now :)
If you LGTM the last commits, I'd merge.

Yes, looks good to me :)
